core/src/test/java/org/elasticsearch/indices/recovery/RecoverySourceHandlerTests.java

```diff
@@ -181,29 +186,74 @@ public void testSendSnapshotSendsOps() throws IOException {
-        final long startingSeqNo = randomBoolean() ? SequenceNumbers.UNASSIGNED_SEQ_NO : randomIntBetween(0, 16);
-        RecoverySourceHandler.SendSnapshotResult result = handler.sendSnapshot(startingSeqNo, new Translog.Snapshot() {
-            @Override
-            public void close() {
+        final long startingSeqNo = randomIntBetween(0, numberOfDocsWithValidSequenceNumbers - 1);
+        final long requiredStartingSeqNo = randomIntBetween((int) startingSeqNo, numberOfDocsWithValidSequenceNumbers - 1);
+        final long endingSeqNo = randomIntBetween((int) requiredStartingSeqNo, numberOfDocsWithValidSequenceNumbers - 1);
+        // todo add proper tests
+        RecoverySourceHandler.SendSnapshotResult result = handler.sendSnapshot(startingSeqNo, requiredStartingSeqNo,
+            endingSeqNo, new Translog.Snapshot() {
+                @Override
+                public void close() {
 
-            }
+                }
 
-            private int counter = 0;
+                private int counter = 0;
 
-            @Override
-            public int totalOperations() {
-                return operations.size() - 1;
-            }
+                @Override
+                public int totalOperations() {
+                    return operations.size() - 1;
+                }
 
-            @Override
-            public Translog.Operation next() throws IOException {
-                return operations.get(counter++);
+                @Override
+                public Translog.Operation next() throws IOException {
+                    return operations.get(counter++);
+                }
+            });
+        final int expectedOps = (int) (endingSeqNo - startingSeqNo + 1);
+        assertThat(result.totalOperations, equalTo(expectedOps));
+        final ArgumentCaptor<List> shippedOpsCaptor = ArgumentCaptor.forClass(List.class);
+        if (expectedOps > 0) {
+            verify(recoveryTarget).indexTranslogOperations(shippedOpsCaptor.capture(), ArgumentCaptor.forClass(Integer.class).capture());
+            List<Translog.Operation> shippedOps = shippedOpsCaptor.getAllValues().stream()
+                .flatMap(List::stream).map(o -> (Translog.Operation) o).collect(Collectors.toList());
+            shippedOps.sort(Comparator.comparing(Translog.Operation::seqNo));
+            assertThat(shippedOps.size(), equalTo(expectedOps));
+            for (int i = 0; i < shippedOps.size(); i++) {
+                assertThat(shippedOps.get(i), equalTo(operations.get(i + (int) startingSeqNo + initialNumberOfDocs)));
             }
-        });
-        if (startingSeqNo == SequenceNumbers.UNASSIGNED_SEQ_NO) {
-            assertThat(result.totalOperations, equalTo(initialNumberOfDocs + numberOfDocsWithValidSequenceNumbers));
         } else {
-            assertThat(result.totalOperations, equalTo(Math.toIntExact(numberOfDocsWithValidSequenceNumbers - startingSeqNo)));
+            verify(recoveryTarget, never()).indexTranslogOperations(null, 0);
```

Comment:
I think this does not verify the right thing. You check that it's never called with the arguments null and 0, whereas you want to check that it's never called with any arguments.

Reply:
good catch. Thanks.